Installation issue: autoconf on ppc64le

### Steps to reproduce the issue

```console
$ spack install autoconf arch=linux-rhel7-power8le
==> Warning: Using GCC 4.8 to optimize for Power 8 might not work if you are not on Red Hat Enterprise Linux 7, where a custom backport of the feature has been done. Upstream support from GCC starts in version 4.9
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/libsigsegv-2.12-txb7lxpsvmwony66zbsjdvp6pvcqgdmi
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/berkeley-db-18.1.40-t3x7wr3zfa6qkvf64wvvgubnltu2tvyl
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/pkgconf-1.7.3-vzmd2eiszaid2mz7cdyufa6a5wtgstmp
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/m4-1.4.18-ujfdwtaemutf3es2doezt4x6rfxiaanm
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/ncurses-6.2-uszrlwxzdwe3nyj2hwhb3jkhtc5lgpkv
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/readline-8.0-djztk7gruh65ryoa4zrvpyiz5rnnmrhz
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/gdbm-1.18.1-2qarpshhhmcc3fozbljikl26t6f5gpls
[+] /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.5/perl-5.32.0-yegtja5ecg5bb36mieemj3q2imtd527l
==> Installing autoconf-2.69-nhgcmyb562pjbsaiie2fokkdielckwwm
==> No binary for autoconf-2.69-nhgcmyb562pjbsaiie2fokkdielckwwm found: installing from source
==> Using cached archive: /home/user/spack-0.16.2/var/spack/cache/_source-cache/archive/95/954bd69b391edc12d6a4a51a2dd1476543da5c6bbf05a95b59dc0dd6fd4c2969.tar.gz
==> autoconf: Executing phase: 'autoreconf'
==> [2021-06-10-12:20:41.891010] CalledProcessError: Command '['/tmp/netuser/spack-stage/spack-stage-autoconf-2.69-nhgcmyb562pjbsaiie2fokkdielckwwm/spack-src/build-aux/config.guess']' returned non-zero exit status 1.
==> Error: RuntimeError: Failed to find suitable substitutes for config.guess

/home/user/spack-0.16.2/lib/spack/spack/build_systems/autotools.py:170, in _do_patch_config_files:
        167
        168        # Check that we found everything we needed
        169        if to_be_found:
  >>    170            msg = 'Failed to find suitable substitutes for {0}'
        171            raise RuntimeError(msg.format(', '.join(to_be_found)))
        172
        173        # Copy the good files over the bad ones
```

### Information on your system

```
spack debug report
* **Spack:** 0.16.2
* **Python:** 3.6.8
* **Platform:** linux-rhel7-power8le
```

### Additional information

This is the same error as in https://github.com/spack/spack/pull/21256, but here the file config.guess (and config.sub) is executable:
```
$ tar -tvf autoconf-2.69.tar.gz |grep config
-rwxrwxr-x 0/0           35516 2012-04-24 19:37 autoconf-2.69/build-aux/config.sub
-rwxrwxr-x 0/0           44826 2012-04-24 19:15 autoconf-2.69/build-aux/config.guess
```

Same error is reported for autoconf 2.62 and 2.59.
Autoconf 2.13 instead is so old that one of the patches fails with:
```
==> Error: FileNotFoundError: [Errno 2] No such file or directory: 'bin/autom4te.in'

/home/user/spack-0.16.2/var/spack/repos/builtin/packages/autoconf/package.py:43, in patch:
         40        # because autom4te is called during the build
         41        filter_file('^#! @PERL@ -w',
         42                    '#! /usr/bin/env perl',
  >>     43                    'bin/autom4te.in')
```

* `spack-build-out.txt` only contains a `CalledProcessError: Command '['/tmp/netuser/spack-stage/spack-stage-autoconf-2.69-nhgcmyb562pjbsaiie2fokkdielckwwm/spack-src/build-aux/config.guess']' returned non-zero exit status 1.`
* [spack-build-env.txt](https://github.com/spack/spack/files/6633907/spack-build-env.txt)

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

I tried manually installing autoconf 2.69, and there's a problem also here:
```
$ ./configure
checking for a BSD-compatible install... /bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking build system type... build-aux/config.guess: unable to guess system type

This script, last modified 2012-02-10, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD
and
  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD

If the version you run (build-aux/config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2012-02-10

uname -m = ppc64le
uname -r = 3.10.0-1127.19.1.el7.ppc64le
uname -s = Linux
uname -v = #1 SMP Tue Aug 11 23:14:00 UTC 2020

/usr/bin/uname -p = ppc64le
/bin/uname -X     =

hostinfo               =
/bin/universe          =
/usr/bin/arch -k       =
/bin/arch              = ppc64le
/usr/bin/oslevel       =
/usr/convex/getsysinfo =

UNAME_MACHINE = ppc64le
UNAME_RELEASE = 3.10.0-1127.19.1.el7.ppc64le
UNAME_SYSTEM  = Linux
UNAME_VERSION = #1 SMP Tue Aug 11 23:14:00 UTC 2020
configure: error: cannot guess build type; you must specify one
```

It seems autoconf ships with an old (2012) version of config.guess, so ppc64 isn't detected.

I've tried adding autoconf 2.71 or 2.70 to spack, but the installation breaks:
```
$ spack install autoconf
==> Installing autoconf-2.71-er4qvp4gejokjmh4l3uo5pok63bomedx
==> No binary for autoconf-2.71-er4qvp4gejokjmh4l3uo5pok63bomedx found: installing from source
==> Fetching file:///home/user/spack-mirror/_source-cache/archive/43/431075ad0bf529ef13cb41e9042c542381103e80015686222b8a9d4abef42a1c.tar.gz
######################################################################## 100.0%
==> autoconf: Executing phase: 'autoreconf'
==> autoconf: Executing phase: 'configure'
==> autoconf: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

4 errors found in build log:
     149    chmod a-w bin/autoreconf.tmp
     150    mv bin/autoscan.tmp bin/autoscan
     151    mv bin/ifnames.tmp bin/ifnames
     152    mv bin/autoreconf.tmp bin/autoreconf
     153    autom4te_perllibdir='/tmp/user/spack-stage/spack-stage-autoconf-2.71-er4qvp4gejokjmh4l3uo5pok63bomedx/
            spack-src'/lib AUTOM4TE_CFG='lib/autom4te.cfg'         bin/autom4te -B ''lib -B '/tmp/user/spack-stage
            /spack-stage-autoconf-2.71-er4qvp4gejokjmh4l3uo5pok63bomedx/spack-src'/lib         --language=M4sh /tmp/n
            etuser/spack-stage/spack-stage-autoconf-2.71-er4qvp4gejokjmh4l3uo5pok63bomedx/spack-src/tests/wrapper.as
            -o tests/wrapper.in
     154    /tmp/user/spack-stage/spack-stage-autoconf-2.71-er4qvp4gejokjmh4l3uo5pok63bomedx/spack-src/build-aux/m
            issing: line 81: help2man: command not found
  >> 155    /bin/sh: bin/autom4te: /home/user/spack-0.16.2/opt/spack/linux-rhel7-power8le/gcc-4.8.: bad
             interpreter: Permission denied
     156    chmod +x bin/autoupdate.tmp
  >> 157    make[1]: *** [bin/autoconf.in] Error 126
     158    make[1]: *** Waiting for unfinished jobs....
     159    chmod a-w bin/autoupdate.tmp
     160    mv bin/autoupdate.tmp bin/autoupdate
     161    WARNING: 'help2man' is missing on your system.
     162             You should only need it if you modified a dependency of a man page.
     163             You may want to install the GNU Help2man package:
     164             <https://www.gnu.org/software/help2man/>
  >> 165    make[1]: *** [man/autom4te.1] Error 127
     166    make[1]: *** wait: No child processes.  Stop.
  >> 167    make: *** [all] Error 2
```

I ended up installing it manually and adding it to `packages.yaml`.

--- a/lib/spack/docs/build_systems/autotoolspackage.rst
+++ b/lib/spack/docs/build_systems/autotoolspackage.rst
@@ -183,6 +183,17 @@ To opt out of this feature, use the following setting:
 
    patch_config_files = False
 
+To enable it conditionally on different architectures, define a property and
+make the package depend on ``gnuconfig`` as a build dependency:
+
+.. code-block
+
+   depends_on('gnuconfig', when='@1.0:')
+
+   @property
+   def patch_config_files(self):
+      return self.spec.satisfies("@1.0:")
+
 .. note::
 
     On some exotic architectures it is necessary to use system provided

--- a/lib/spack/docs/build_systems/autotoolspackage.rst
+++ b/lib/spack/docs/build_systems/autotoolspackage.rst
@@ -159,6 +159,46 @@ create a new patch that directly modifies ``configure``. That way,
 Spack can use the secondary patch and additional build system
 dependencies aren't necessary.
 
+""""""""""""""""""""""""""""
+Old Autotools helper scripts
+""""""""""""""""""""""""""""
+
+Autotools based tarballs come with helper scripts such as ``config.sub`` and
+``config.guess``. It is the responsibility of the developers to keep these files
+up to date so that they run on every platform, but for very old software
+releases this is impossible. In these cases Spack can help to replace these
+files with newer ones, without having to add the heavy dependency on
+``automake``.
+
+Automatic helper script replacement is currently enabled by default on
+``ppc64le`` and ``aarch64``, as these are the known cases where old scripts fail.
+On these targets, ``AutotoolsPackage`` adds a build dependency on ``gnuconfig``,
+which is a very light-weight package with newer versions of the helper files.
+Spack then tries to run all the helper scripts it can find in the release, and
+replaces them on failure with the helper scripts from ``gnuconfig``.
+
+To opt out of this feature, use the following setting:
+
+.. code-block:: python
+
+   patch_config_files = False
+
+.. note::
+
+    On some exotic architectures it is necessary to use system provided
+    ``config.sub`` and ``config.guess`` files. In this case, the most
+    transparent solution is to mark the ``gnuconfig`` package as external and
+    non-buildable, with a prefix set to the directory containing the files:
+
+   .. code-block:: yaml
+
+       gnuconfig:
+         buildable: false
+         externals:
+         - spec: gnuconfig@master
+           prefix: /usr/share/configure_files/
+
+
 """"""""""""""""
 force_autoreconf
 """"""""""""""""

--- a/lib/spack/spack/build_systems/autotools.py
+++ b/lib/spack/spack/build_systems/autotools.py
@@ -55,15 +55,22 @@ class AutotoolsPackage(PackageBase):
     #: This attribute is used in UI queries that need to know the build
     #: system base class
     build_system_class = 'AutotoolsPackage'
-    #: Whether or not to update old ``config.guess`` and ``config.sub`` files
-    #: distributed with the tarball. This currently only applies to ``ppc64le:``
-    #: and ``aarch64:`` target architectures. The substitutes are taken from the
-    #: ``gnuconfig`` package, which is automatically added as a build dependency
-    #: for these architectures. In case system versions of these config files are
-    #: required, the ``gnuconfig`` package can be marked external with a prefix
-    #: pointing to the directory containing the system ``config.guess`` and
-    #: ``config.sub`` files.
-    patch_config_files = True
+
+    @property
+    def patch_config_files(self):
+        """
+        Whether or not to update old ``config.guess`` and ``config.sub`` files
+        distributed with the tarball. This currently only applies to ``ppc64le:``
+        and ``aarch64:`` target architectures. The substitutes are taken from the
+        ``gnuconfig`` package, which is automatically added as a build dependency
+        for these architectures. In case system versions of these config files are
+        required, the ``gnuconfig`` package can be marked external with a prefix
+        pointing to the directory containing the system ``config.guess`` and
+        ``config.sub`` files.
+        """
+        return (self.spec.satisfies('target=ppc64le:')
+                or self.spec.satisfies('target=aarch64:'))
+
     #: Whether or not to update ``libtool``
     #: (currently only for Arm/Clang/Fujitsu compilers)
     patch_libtool = True
@@ -116,18 +123,15 @@ class AutotoolsPackage(PackageBase):
         In particular, config.guess fails for PPC64LE for version prior
         to a 2013-06-10 build date (automake 1.13.4) and for ARM (aarch64).
         """
-        is_ppc64le = self.spec.satisfies('target=ppc64le:')
-        is_aarch64 = self.spec.satisfies('target=aarch64:')
-
-        if not self.patch_config_files or (not is_ppc64le and not is_aarch64):
+        if not self.patch_config_files:
             return
 
         # TODO: Expand this to select the 'config.sub'-compatible architecture
         # for each platform (e.g. 'config.sub' doesn't accept 'power9le', but
         # does accept 'ppc64le').
-        if is_ppc64le:
+        if self.spec.satisfies('target=ppc64le:'):
             config_arch = 'ppc64le'
-        elif is_aarch64:
+        elif self.spec.satisfies('target=aarch64:'):
             config_arch = 'aarch64'
         else:
             config_arch = 'local'
@@ -156,6 +160,12 @@ class AutotoolsPackage(PackageBase):
         if not to_be_patched:
             return
 
+        # Otherwise, require `gnuconfig` to be a build dependency
+        self._require_build_deps(
+            pkgs=['gnuconfig'],
+            spec=self.spec,
+            err="Cannot patch config files")
+
         # Get the config files we need to patch (config.sub / config.guess).
         to_be_found = list(set(os.path.basename(f) for f in to_be_patched))
         gnuconfig = self.spec['gnuconfig']
@@ -173,18 +183,12 @@ class AutotoolsPackage(PackageBase):
         # For external packages the user may have specified an incorrect prefix.
         # otherwise the installation is just corrupt.
         if not candidates:
+            msg = ("Spack could not find `config.guess` and `config.sub` "
+                   "files in the `gnuconfig` prefix `{0}`. This means the "
+                   "`gnuconfig` package is broken").format(gnuconfig_dir)
             if gnuconfig.external:
-                msg = ("Spack could not find `config.guess` and `config.sub` "
-                       "files in the `gnuconfig` prefix `{0}`. This means you "
-                       "have set an incorrect prefix path for the external "
-                       "`gnuconfig` package. Make sure the prefix is correct, or "
-                       "remove `gnuconfig` as an external "
-                       "package.").format(gnuconfig_dir)
-            else:
-                msg = ("Spack could not find `config.guess` and `config.sub` "
-                       "files in the `gnuconfig` prefix `{0}`. This means the "
-                       "`gnuconfig` package is corrupt. Please file an "
-                       "issue.").format(gnuconfig_dir)
+                msg += (" or the `gnuconfig` package prefix is misconfigured as"
+                        " an external package")
             raise InstallError(msg)
 
         # Filter working substitutes
@@ -292,30 +296,40 @@ To resolve this problem, please try the following:
         if self.force_autoreconf:
             force_remove(self.configure_abs_path)
 
-    def _autoreconf_warning(self, spec, missing):
-        msg = ("Cannot generate configure: missing dependencies {0}.\n\nPlease add "
-               "the following lines to the package:\n\n".format(", ".join(missing)))
+    def _require_build_deps(self, pkgs, spec, err):
+        """Require `pkgs` to be direct build dependencies of `spec`. Raises a
+        RuntimeError with a helpful error messages when any dep is missing."""
 
-        for dep in missing:
+        build_deps = [d.name for d in spec.dependencies(deptype='build')]
+        missing_deps = [x for x in pkgs if x not in build_deps]
+
+        if not missing_deps:
+            return
+
+        # Raise an exception on missing deps.
+        msg = ("{0}: missing dependencies: {1}.\n\nPlease add "
+               "the following lines to the package:\n\n"
+               .format(err, ", ".join(missing_deps)))
+
+        for dep in missing_deps:
             msg += ("    depends_on('{0}', type='build', when='@{1}')\n"
                     .format(dep, spec.version))
 
         msg += "\nUpdate the version (when='@{0}') as needed.".format(spec.version)
-
-        return msg
+        raise RuntimeError(msg)
 
     def autoreconf(self, spec, prefix):
         """Not needed usually, configure should be already there"""
+
         # If configure exists nothing needs to be done
         if os.path.exists(self.configure_abs_path):
             return
-        # Else try to regenerate it
-        needed_dependencies = ['autoconf', 'automake', 'libtool']
-        build_deps = [d.name for d in spec.dependencies(deptype='build')]
-        missing = [x for x in needed_dependencies if x not in build_deps]
 
-        if missing:
-            raise RuntimeError(self._autoreconf_warning(spec, missing))
+        # Else try to regenerate it, which reuquires a few build dependencies
+        self._require_build_deps(
+            pkgs=['autoconf', 'automake', 'libtool'],
+            spec=spec,
+            err="Cannot generate configure")
 
         tty.msg('Configure script not found: trying to generate it')
         tty.warn('*********************************************************')

--- a/lib/spack/spack/build_systems/autotools.py
+++ b/lib/spack/spack/build_systems/autotools.py
@@ -3,7 +3,6 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 import inspect
-import itertools
 import os
 import os.path
 import stat
@@ -14,6 +13,8 @@ import llnl.util.filesystem as fs
 import llnl.util.tty as tty
 from llnl.util.filesystem import force_remove, working_dir
 
+from spack.build_environment import InstallError
+from spack.directives import depends_on
 from spack.package import PackageBase, run_after, run_before
 from spack.util.executable import Executable
 
@@ -54,8 +55,14 @@ class AutotoolsPackage(PackageBase):
     #: This attribute is used in UI queries that need to know the build
     #: system base class
     build_system_class = 'AutotoolsPackage'
-    #: Whether or not to update ``config.guess`` and ``config.sub`` on old
-    #: architectures
+    #: Whether or not to update old ``config.guess`` and ``config.sub`` files
+    #: distributed with the tarball. This currently only applies to ``ppc64le:``
+    #: and ``aarch64:`` target architectures. The substitutes are taken from the
+    #: ``gnuconfig`` package, which is automatically added as a build dependency
+    #: for these architectures. In case system versions of these config files are
+    #: required, the ``gnuconfig`` package can be marked external with a prefix
+    #: pointing to the directory containing the system ``config.guess`` and
+    #: ``config.sub`` files.
     patch_config_files = True
     #: Whether or not to update ``libtool``
     #: (currently only for Arm/Clang/Fujitsu compilers)
@@ -83,6 +90,9 @@ class AutotoolsPackage(PackageBase):
     #: after the installation. If True instead it installs them.
     install_libtool_archives = False
 
+    depends_on('gnuconfig', type='build', when='target=ppc64le:')
+    depends_on('gnuconfig', type='build', when='target=aarch64:')
+
     @property
     def _removed_la_files_log(self):
         """File containing the list of remove libtool archives"""
@@ -106,18 +116,18 @@ class AutotoolsPackage(PackageBase):
         In particular, config.guess fails for PPC64LE for version prior
         to a 2013-06-10 build date (automake 1.13.4) and for ARM (aarch64).
         """
-        if not self.patch_config_files or (
-                not self.spec.satisfies('target=ppc64le:') and
-                not self.spec.satisfies('target=aarch64:')
-        ):
+        is_ppc64le = self.spec.satisfies('target=ppc64le:')
+        is_aarch64 = self.spec.satisfies('target=aarch64:')
+
+        if not self.patch_config_files or (not is_ppc64le and not is_aarch64):
             return
 
         # TODO: Expand this to select the 'config.sub'-compatible architecture
         # for each platform (e.g. 'config.sub' doesn't accept 'power9le', but
         # does accept 'ppc64le').
-        if self.spec.satisfies('target=ppc64le:'):
+        if is_ppc64le:
             config_arch = 'ppc64le'
-        elif self.spec.satisfies('target=aarch64:'):
+        elif is_aarch64:
             config_arch = 'aarch64'
         else:
             config_arch = 'local'
@@ -138,39 +148,69 @@ class AutotoolsPackage(PackageBase):
 
             return True
 
-        # Compute the list of files that needs to be patched
-        search_dir = self.stage.path
-        to_be_patched = fs.find(
-            search_dir, files=['config.sub', 'config.guess'], recursive=True
-        )
+        # Get the list of files that needs to be patched
+        to_be_patched = fs.find(self.stage.path, files=['config.sub', 'config.guess'])
         to_be_patched = [f for f in to_be_patched if not runs_ok(f)]
 
         # If there are no files to be patched, return early
         if not to_be_patched:
             return
 
-        # Directories where to search for files to be copied
-        # over the failing ones
-        good_file_dirs = ['/usr/share']
-        if 'automake' in self.spec:
-            good_file_dirs.insert(0, self.spec['automake'].prefix)
-
-        # List of files to be found in the directories above
+        # Get the config files we need to patch (config.sub / config.guess).
         to_be_found = list(set(os.path.basename(f) for f in to_be_patched))
+        gnuconfig = self.spec['gnuconfig']
+        gnuconfig_dir = gnuconfig.prefix
+
+        # An external gnuconfig may not not have a prefix.
+        if gnuconfig_dir is None:
+            raise InstallError("Spack could not find substitutes for GNU config "
+                               "files because no prefix is available for the "
+                               "`gnuconfig` package. Make sure you set a prefix "
+                               "path instead of modules for external `gnuconfig`.")
+
+        candidates = fs.find(gnuconfig_dir, files=to_be_found, recursive=False)
+
+        # For external packages the user may have specified an incorrect prefix.
+        # otherwise the installation is just corrupt.
+        if not candidates:
+            if gnuconfig.external:
+                msg = ("Spack could not find `config.guess` and `config.sub` "
+                       "files in the `gnuconfig` prefix `{0}`. This means you "
+                       "have set an incorrect prefix path for the external "
+                       "`gnuconfig` package. Make sure the prefix is correct, or "
+                       "remove `gnuconfig` as an external "
+                       "package.").format(gnuconfig_dir)
+            else:
+                msg = ("Spack could not find `config.guess` and `config.sub` "
+                       "files in the `gnuconfig` prefix `{0}`. This means the "
+                       "`gnuconfig` package is corrupt. Please file an "
+                       "issue.").format(gnuconfig_dir)
+            raise InstallError(msg)
+
+        # Filter working substitutes
+        candidates = [f for f in candidates if runs_ok(f)]
         substitutes = {}
-        for directory in good_file_dirs:
-            candidates = fs.find(directory, files=to_be_found, recursive=True)
-            candidates = [f for f in candidates if runs_ok(f)]
-            for name, good_files in itertools.groupby(
-                    candidates, key=os.path.basename
-            ):
-                substitutes[name] = next(good_files)
-                to_be_found.remove(name)
+        for candidate in candidates:
+            config_file = os.path.basename(candidate)
+            substitutes[config_file] = candidate
+            to_be_found.remove(config_file)
 
         # Check that we found everything we needed
         if to_be_found:
-            msg = 'Failed to find suitable substitutes for {0}'
-            raise RuntimeError(msg.format(', '.join(to_be_found)))
+            msg = """\
+Spack could not find working replacements for the following autotools config
+files: {0}.
+
+To resolve this problem, please try the following:
+1. Try to rebuild with `patch_config_files = False` in the package `{1}`, to
+   rule out that Spack tries to replace config files not used by the build.
+2. Verify that the `gnuconfig` package is up-to-date.
+3. On some systems you need to use system-provided `config.guess` and `config.sub`
+   files. In this case, mark `gnuconfig` as an non-buildable external package,
+   and set the prefix to the directory containing the `config.guess` and
+   `config.sub` files.
+"""
+            raise InstallError(msg.format(', '.join(to_be_found), self.name))
 
         # Copy the good files over the bad ones
         for abs_path in to_be_patched:

--- a/lib/spack/spack/test/build_systems.py
+++ b/lib/spack/spack/test/build_systems.py
@@ -10,8 +10,10 @@ import pytest
 
 import llnl.util.filesystem as fs
 
+import spack.architecture
+import spack.environment
 import spack.repo
-from spack.build_environment import get_std_cmake_args, setup_package
+from spack.build_environment import ChildError, get_std_cmake_args, setup_package
 from spack.spec import Spec
 from spack.util.executable import which
 
@@ -190,7 +192,7 @@ class TestAutotoolsPackage(object):
             self, mutable_database
     ):
         # Install a package that creates a mock libtool archive
-        s = spack.spec.Spec('libtool-deletion')
+        s = Spec('libtool-deletion')
         s.concretize()
         s.package.do_install(explicit=True)
 
@@ -208,13 +210,96 @@ class TestAutotoolsPackage(object):
     ):
         # Install a package that creates a mock libtool archive,
         # patch its package to preserve the installation
-        s = spack.spec.Spec('libtool-deletion')
+        s = Spec('libtool-deletion')
         s.concretize()
         monkeypatch.setattr(s.package, 'install_libtool_archives', True)
         s.package.do_install(explicit=True)
 
         # Assert libtool archives are installed
         assert os.path.exists(s.package.libtool_archive_file)
+
+    def test_autotools_gnuconfig_replacement(self, mutable_database):
+        """
+        Tests whether only broken config.sub and config.guess are replaced with
+        files from working alternatives from the gnuconfig package.
+        """
+        s = Spec('autotools-config-replacement +patch_config_files +gnuconfig')
+        s.concretize()
+        s.package.do_install()
+
+        with open(os.path.join(s.prefix.broken, 'config.sub')) as f:
+            assert "gnuconfig version of config.sub" in f.read()
+
+        with open(os.path.join(s.prefix.broken, 'config.guess')) as f:
+            assert "gnuconfig version of config.guess" in f.read()
+
+        with open(os.path.join(s.prefix.working, 'config.sub')) as f:
+            assert "gnuconfig version of config.sub" not in f.read()
+
+        with open(os.path.join(s.prefix.working, 'config.guess')) as f:
+            assert "gnuconfig version of config.guess" not in f.read()
+
+    def test_autotools_gnuconfig_replacement_disabled(self, mutable_database):
+        """
+        Tests whether disabling patch_config_files
+        """
+        s = Spec('autotools-config-replacement ~patch_config_files +gnuconfig')
+        s.concretize()
+        s.package.do_install()
+
+        with open(os.path.join(s.prefix.broken, 'config.sub')) as f:
+            assert "gnuconfig version of config.sub" not in f.read()
+
+        with open(os.path.join(s.prefix.broken, 'config.guess')) as f:
+            assert "gnuconfig version of config.guess" not in f.read()
+
+        with open(os.path.join(s.prefix.working, 'config.sub')) as f:
+            assert "gnuconfig version of config.sub" not in f.read()
+
+        with open(os.path.join(s.prefix.working, 'config.guess')) as f:
+            assert "gnuconfig version of config.guess" not in f.read()
+
+    @pytest.mark.disable_clean_stage_check
+    def test_autotools_gnuconfig_replacement_no_gnuconfig(self, mutable_database):
+        """
+        Tests whether a useful error message is shown when patch_config_files is
+        enabled, but gnuconfig is not listed as a direct build dependency.
+        """
+        s = Spec('autotools-config-replacement +patch_config_files ~gnuconfig')
+        s.concretize()
+
+        msg = "Cannot patch config files: missing dependencies: gnuconfig"
+        with pytest.raises(ChildError, match=msg):
+            s.package.do_install()
+
+    @pytest.mark.disable_clean_stage_check
+    def test_broken_external_gnuconfig(self, mutable_database, tmpdir):
+        """
+        Tests whether we get a useful error message when gnuconfig is marked
+        external, but the install prefix is misconfigured and no config.guess
+        and config.sub substitute files are found in the provided prefix.
+        """
+        env_dir = str(tmpdir.ensure('env', dir=True))
+        gnuconfig_dir = str(tmpdir.ensure('gnuconfig', dir=True))  # empty dir
+        with open(os.path.join(env_dir, 'spack.yaml'), 'w') as f:
+            f.write("""\
+spack:
+  specs:
+  - 'autotools-config-replacement +patch_config_files +gnuconfig'
+  packages:
+    gnuconfig:
+      buildable: false
+      externals:
+      - spec: gnuconfig@1.0.0
+        prefix: {0}
+""".format(gnuconfig_dir))
+
+        msg = ("Spack could not find `config.guess`.*misconfigured as an "
+               "external package")
+        with spack.environment.Environment(env_dir) as e:
+            e.concretize()
+            with pytest.raises(ChildError, match=msg):
+                e.install_all()
 
 
 @pytest.mark.usefixtures('config', 'mock_packages')

--- a/var/spack/repos/builtin.mock/packages/autotools-config-replacement/package.py
+++ b/var/spack/repos/builtin.mock/packages/autotools-config-replacement/package.py
@@ -61,13 +61,13 @@ class AutotoolsConfigReplacement(AutotoolsPackage):
         # a configure script is required
         configure_script = join_path(self.stage.source_path, 'configure')
         with open(configure_script, 'w') as f:
-            f.write("#!/bin/sh -c 'exit 1'")
+            f.write("#!/bin/sh\nexit 0")
         os.chmod(configure_script, 0o775)
 
         # broken config.sub (not executable)
         broken_config_sub = join_path(broken, 'config.sub')
         with open(broken_config_sub, 'w') as f:
-            f.write("#!/bin/sh")
+            f.write("#!/bin/sh\nexit 0")
 
         # broken config.guess (exectuable but with error return code)
         broken_config_guess = join_path(broken, 'config.guess')
@@ -78,11 +78,11 @@ class AutotoolsConfigReplacement(AutotoolsPackage):
         # working config.sub
         working_config_sub = join_path(working, 'config.sub')
         with open(working_config_sub, 'w') as f:
-            f.write("#!/bin/sh")
+            f.write("#!/bin/sh\nexit 0")
         os.chmod(working_config_sub, 0o775)
 
         # working config.guess
         working_config_guess = join_path(working, 'config.guess')
         with open(working_config_guess, 'w') as f:
-            f.write("#!/bin/sh")
+            f.write("#!/bin/sh\nexit 0")
         os.chmod(working_config_guess, 0o775)

--- a/var/spack/repos/builtin.mock/packages/autotools-config-replacement/package.py
+++ b/var/spack/repos/builtin.mock/packages/autotools-config-replacement/package.py
@@ -1,0 +1,88 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+from spack import *
+
+
+class AutotoolsConfigReplacement(AutotoolsPackage):
+    """
+    This package features broken and working config.sub and config.guess files,
+    that should be replaced by the ones provided by gnuconfig. It allows testing
+    with / without patches and with / without substitutes available.
+    """
+
+    has_code = False
+
+    version('1.0.0')
+    variant('patch_config_files', default=False)
+    variant('gnuconfig', default=False)
+
+    depends_on('gnuconfig', type='build', when='+gnuconfig')
+
+    @property
+    def patch_config_files(self):
+        return self.spec.satisfies('+patch_config_files')
+
+    def autoreconf(self, spec, prefix):
+        pass
+
+    def configure(self, spec, prefix):
+        pass
+
+    def build(self, spec, prefix):
+        pass
+
+    def install(self, spec, prefix):
+        broken = os.path.join(self.stage.source_path, 'broken')
+        working = os.path.join(self.stage.source_path, 'working')
+        install_tree(broken, self.prefix.broken)
+        install_tree(working, self.prefix.working)
+
+    @run_before('autoreconf')
+    def create_the_package_sources(self):
+        # Creates the following file structure:
+        # ./broken/config.sub    -- not executable
+        # ./broken/config.guess  -- exectuable & exit code 1
+        # ./working/config.sub   -- executable & exit code 0
+        # ./working/config.guess -- executable & exit code 0
+        # Automatic config helper script substitution should replace the two
+        # broken scripts with those from the gnuconfig package.
+
+        broken = os.path.join(self.stage.source_path, 'broken')
+        working = os.path.join(self.stage.source_path, 'working')
+
+        mkdirp(broken)
+        mkdirp(working)
+
+        # a configure script is required
+        configure_script = join_path(self.stage.source_path, 'configure')
+        with open(configure_script, 'w') as f:
+            f.write("#!/bin/sh -c 'exit 1'")
+        os.chmod(configure_script, 0o775)
+
+        # broken config.sub (not executable)
+        broken_config_sub = join_path(broken, 'config.sub')
+        with open(broken_config_sub, 'w') as f:
+            f.write("#!/bin/sh")
+
+        # broken config.guess (exectuable but with error return code)
+        broken_config_guess = join_path(broken, 'config.guess')
+        with open(broken_config_guess, 'w') as f:
+            f.write("#!/bin/sh\nexit 1")
+        os.chmod(broken_config_guess, 0o775)
+
+        # working config.sub
+        working_config_sub = join_path(working, 'config.sub')
+        with open(working_config_sub, 'w') as f:
+            f.write("#!/bin/sh")
+        os.chmod(working_config_sub, 0o775)
+
+        # working config.guess
+        working_config_guess = join_path(working, 'config.guess')
+        with open(working_config_guess, 'w') as f:
+            f.write("#!/bin/sh")
+        os.chmod(working_config_guess, 0o775)

--- a/var/spack/repos/builtin.mock/packages/gnuconfig/package.py
+++ b/var/spack/repos/builtin.mock/packages/gnuconfig/package.py
@@ -3,6 +3,8 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+import os
+
 from spack import *
 
 
@@ -18,5 +20,16 @@ class Gnuconfig(Package):
     version('2021-08-14')
 
     def install(self, spec, prefix):
-        touch(join_path(prefix, 'config.sub'))
-        touch(join_path(prefix, 'config.guess'))
+        config_sub = join_path(prefix, 'config.sub')
+        config_guess = join_path(prefix, 'config.guess')
+
+        # Create files
+        with open(config_sub, 'w') as f:
+            f.write("#!/bin/sh\necho gnuconfig version of config.sub")
+
+        with open(config_guess, 'w') as f:
+            f.write("#!/bin/sh\necho gnuconfig version of config.guess")
+
+        # Make executable
+        os.chmod(config_sub, 0o775)
+        os.chmod(config_guess, 0o775)

--- a/var/spack/repos/builtin.mock/packages/gnuconfig/package.py
+++ b/var/spack/repos/builtin.mock/packages/gnuconfig/package.py
@@ -1,0 +1,22 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Gnuconfig(Package):
+    """
+    The GNU config.guess and config.sub scripts versioned by timestamp.
+    This package can be used as a build dependency for autotools packages that
+    ship a tarball with outdated config.guess and config.sub files.
+    """
+
+    has_code = False
+
+    version('2021-08-14')
+
+    def install(self, spec, prefix):
+        touch(join_path(prefix, 'config.sub'))
+        touch(join_path(prefix, 'config.guess'))

